### PR TITLE
pds-api 221: rework all/latest

### PR DESCRIPTION
**Pull request by al-niessner**
## 🗒️ Summary
Solution ended at Plan C. There is a new ops property "ops:Provenance/ops:superseded_by" that has been [added to the registry index](https://github.com/NASA-PDS/registry-mgr/pull/63) but waiting to integrated. When a product become part of provenance (not latest) "ops:Provenance/ops:superseded_by" is set to the new `lidvid`. When searches via registry-api are performed, it filters on the existence of "ops:Provenance/ops:superseded_by" where latest it does not exist.

The script [support/provenance.py](https://github.com/NASA-PDS/registry-api/blob/pds-api-221/support/provenance.py) does the work of building the history. It takes 11 minutes in current cluster configuration for 2 million lidvids and 15 minutes for 6 million. 

Future of support/provenance.py would be to mix it with changing the tracking status from staged to not-staged. While moving from an internal state to available to the outside world state, the provenance should also be updated. It may possible to then optimize the provenance algorithm to reduce processing time. The trade between current algorithmic approach and an optimized approach is robustness of catching changes made outside of a standard script for moving out of staged.

## ⚙️ Test Data and/or Report

Required adding historical data then running support/provenance.py before these tests would be meaningful.

```
$ curl -X GET 'http://localhost:8080/products/urn:nasa:pds:insight_rad:data_calibrated?fields=lidvid,ops:Provenance.ops:superseded_by&limit=30' --header 'Accept: application/kvp+json'
{
  "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::7.0"
}

$ curl -X GET 'http://localhost:8080/products/urn:nasa:pds:insight_rad:data_calibrated/all?fields=lidvid,ops:Provenance.ops:superseded_by&limit=30' --header 'Accept: application/kvp+json'
{
  "summary":{"q":"","hits":37,"took":18,"start":0,"limit":30,"sort":[],"properties":["lidvid","ops:Provenance.ops:superseded_by"]},
  "data":[    {
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::7.0"    },
    {
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::6.5"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::6.5",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::6.4"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::6.4",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::6.3"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::6.3",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::6.2"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::6.2",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::6.1"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::6.1",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::6.0"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::6.0",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::5.5"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::5.5",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::5.4"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::5.4",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::5.3"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::5.3",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::5.2"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::5.2",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::5.1"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::5.1",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::5.0"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::5.0",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::4.5"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::4.5",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::4.4"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::4.4",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::4.3"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::4.3",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::4.2"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::4.2",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::4.1"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::4.1",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::4.0"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::4.0",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::3.5"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::3.5",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::3.4"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::3.4",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::3.3"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::3.3",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::3.2"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::3.2",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::3.1"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::3.1",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::3.0"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::3.0",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::2.5"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::2.5",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::2.4"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::2.4",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::2.3"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::2.3",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::2.2"    },
    {
      "ops:Provenance.ops:superseded_by":"urn:nasa:pds:insight_rad:data_calibrated::2.2",
      "lidvid":"urn:nasa:pds:insight_rad:data_calibrated::2.1"    }
  ]
}
```

## ♻️ Related Issues

- NASA-PDS/pdS-api#220
- NASA-PDS/registry-api#441

### CODE REVIEW

File: support/update_latest.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    return datetime.datetime.fromisoformat (input)
+
+def _log_level (input:str)->int:
+    try: result = int(input)
+    except ValueError: result = getattr (logging, input)
+    return result
+
+def add_doc (reg, doc, id):
+    response = requests.put ('https://localhost:9200/' + reg + '/_doc/' + id,
+                             auth=('admin', 'admin'), json=doc, verify=False)
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/update_latest.py]: ignore

File: support/update_latest.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+    for lidvid in troll_registry (host, args.time_begin, args.time_end):
+        log.info ('processing %s', lidvid)
+        latest = get_latest (host, lidvid.split('::')[0])
+
+        if is_newer (latest, lidvid): update (host, latest, lidvid)
+        pass
+    return
+
+def create_latest_index (host:HOST, config:{}):
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/update_latest.py]: ignore

File: support/harvest_range.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            more_data = False
+            log.error ('Bad response code (%d): %s',
+                       resp.status_code, resp.reason)
+            pass
+        pass
+
+    if 'scroll_id' in query:
+        path = '_search/scroll/' + query['scroll_id']
+        requests.delete (urllib.parse.urljoin (host.url, path),
+                         auth=(host.username, host.password), verify=False)
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/harvest_range.py]: ignore

File: support/harvest_range.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            more_data = total < data['hits']['total']['value']
+        else:
+            more_data = False
+            log.error ('Bad response code (%d): %s',
+                       resp.status_code, resp.reason)
+            pass
+        pass
+
+    if 'scroll_id' in query:
+        path = '_search/scroll/' + query['scroll_id']
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/harvest_range.py]: ignore

File: support/harvest_range.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    more_data = True
+    path = 'registry/_search?scroll=10m'
+    query = {'query':{'match_all':{}},
+             '_source':{'includes':['ops:Harvest_Info/ops:harvest_date_time']},
+             'size':5 }
+    total = 0
+    while more_data:
+        resp = requests.get (urllib.parse.urljoin (host.url, path),
+                             auth=(host.username, host.password),
+                             verify=False, json=query)
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/harvest_range.py]: ignore

File: support/harvest_range.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/harvest_range.py]: ignore

File: support/update_latest.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+        log.warning ('latest index contained more versions than latest')
+        clean_latest (host, discard)
+        pass
+
+    reindex_lidvids,discard = merge_indices (lidvids, latest_lidvids)
+    clean_latest (host, discard)
+    reindex (host, reindex_lidvids)
+    return
+
+def clean_latest (host:HOST, lidvids:[str]):
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/update_latest.py]: ignore

File: support/update_latest.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        bulk = '\n'.join ([json.dumps({'delete':{'_id':lidvid}}) for lidvid in lidvids]) + '\n'
+        headers = {'Content-Type': 'application/x-ndjson'}
+        response = requests.put (urllib.parse.urljoin (host.url,'latest/_bulk'),
+                                 auth=(host.username, host.password),
+                                 data=bulk, headers=headers,
+                                 verify=host.verify)
+
+        if response.status_code != 200:
+            log.error ('could not clean latest index because of error (%d): %s',
+                       response.status_code, response,reason)
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/update_latest.py]: ignore

File: support/provenance.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                        format='%(asctime)s::%(levelname)s::%(message)s')
+    log.info ('starting CLI processing')
+    host = HOST(args.cluster_nodes, args.password, args.base_URL, args.username,
+                args.verify)
+    provenance = troll_registry (host)
+    updates = get_historic (provenance, args.reset)
+    if updates: update_docs (host, updates)
+    return
+
+def get_historic (provenance:{str:str}, reset:bool)->{str:str}:
```
[the remark this answers is not in the document]
Comment (al-niessner) [reply support/provenance.py:73]: ignore

**Comment by al-niessner** [comment]
@jordanpadams @tloubrieu-jpl 

Ready for review. Including @alexdunnjpl as they seem to be working in this repo too.

@jimmie there is the pull request NASA-PDS/registry-mgr#63 that needs to moved through as well as the live system will need to be reindexed with the field. I thought I could postpone it with a post filter but opensearch ditched that concept - it was/is an elasticsearch concept. It does not need to be done this second or the world blows up, but it has to be completed for the end user to see this problem as being fixed. opensearch only searches those items in the index.

**Comment by jordanpadams** [comment]
@al-niessner would you mind updating the primary description for this PR reflect the change in the overall design approach that has been implemented.

additionally, @tloubrieu-jpl @al-niessner where did we land on defining a new `ops.Tracking_Meta.ops.tracking_status` (or something like that) for handling how harvest and registry manager should actually be determining "staged" vs. "in_queue_release" vs. etc.? should I create another ticket for that? I think that is really needed in order to track those intermediate states before the `provenance.py` script is executed.

also, do we want to maybe move `provenance.py` to some other `registry-api-scripts` repo (or maybe the [pds-api](https://github.com/NASA-PDS/pds-api) repo)? that way it can be an installable Python package we can run more easily as a separate tool? just a thought.

**Comment by tloubrieu-jpl** [comment]
Hi @jordanpadams it is on my tasks for this week to wrap this topics and make sure we have all the documentation for users, administrator, and developers.

**Comment by jordanpadams** [comment]
@tloubrieu-jpl great! let me know if there is anything I can help with

**Review by tloubrieu-jpl — APPROVED** [review]
Good to go, thanks @al-niessner

**Pull request merged**